hysds/utils.py

```diff
@@ -343,15 +349,29 @@ def get_payload_hash(payload):
     ).hexdigest()
 
 
+def no_dedup_job(details):
+    logger.info("Giving up querying for dedup jobs with args {args} and kwargs {kwargs}".format(**details))
+    return None
+
+
 @backoff.on_exception(
     backoff.expo, requests.exceptions.RequestException, max_tries=8, max_value=32
 )
+@backoff.on_exception(
+    backoff.expo, NoDedupJobFoundException, max_tries=8, max_value=32, on_giveup=no_dedup_job
+)
 def query_dedup_job(dedup_key, filter_id=None, states=None):
     """
     Return job IDs with matching dedup key defined in states
     'job-queued', 'job-started', 'job-completed', by default.
     """
 
+    hash_exists_in_redis = payload_hash_exists(dedup_key)
```

Comment:
@mcayanan should we move this to line 418 inside this `if` statement?
```python
if j["hits"]["total"]["value"] == 0:
```
it can save a little bit of time of having to check redis `payload_hash` for every job

Reply:
@DustinKLo: We need that atomic redis check up front to resolve the race condition between two duplicate jobs starting around the same time. Without that, they both will query ES and get the same result: that there are no other jobs in the system with that same payload hash.

Reply:
but it looks like we're not doing anything with the `hash_exists_in_redis` value until after we query elsaticsearch and nothing shows up?

Reply:
When `hash_exists_in_redis` is True, that means we have identified that there is a possible duplicate job in the system. That's why in line 419, we do a check if this flag is True or not. If it is True, then we need to sort of "wait" and query ES a couple of times to give it time for the duplicate job to make it into the system.